contrib/dhcp_agent/dhcp_agent.py

```diff
@@ -0,0 +1,399 @@
+        )
+        return False
+    return True
+
+
+class DHCPConfigManager(object):
+    def __init__(
+        self, logger, host, key, dc=None, net_env=None, mode=MODE_ALL,
+        verbose=False, restart=False, proto=PROTO_HTTPS,
+        dhcp_config_entries='', dhcp_config_networks='',
+        dhcp_service_name=DEFAULT_DHCP_SERVICE_NAME, **kwargs
+    ):
+        self.load_config()
+        self.logger = logger
+        self.host = host
+        self.key = key
+        self.proto = proto
+        self.can_restart_dhcp_server = restart
+        self.dhcp_service_name = dhcp_service_name
+        self.dhcp_entries_config_path = dhcp_config_entries
+        self.dhcp_networks_config_path = dhcp_config_networks
+        self.envs = net_env.split(',') if net_env else []
+        self.dcs = dc.split(',') if dc else []
+        self.mode = mode
+
+    def download_and_apply_configuration(self):
+        request_params = self.get_request_params(self.dcs, self.envs)
+        should_restart_dhcp_server = False
+        if self.mode == MODE_ALL:
+            fetch_statuses = []
+            for m in MODES_EXCLUDE_ALL:
+                fetch_statuses.append(self.fetch_configuration(
+                    m, request_params
+                ))
+            should_restart_dhcp_server = any(fetch_statuses)
+        else:
+            should_restart_dhcp_server = self.fetch_configuration(
+                self.mode, request_params
+            )
+
+        if self.can_restart_dhcp_server and should_restart_dhcp_server:
+            self._restart_dhcp_server()
+        self._send_synch_confirmation()
+        self.save_config()
+
+    def make_request(self, url, data=None):
+        data = urlencode(data).encode('utf-8') if data else None
+        headers = {}
+        last = self.get_last_modified(url)
+        if last:
+            headers['If-Modified-Since'] = last
+        headers.update({'Authorization': 'Token {}'.format(self.key)})
+        return urlopen(Request(url, data=data, headers=headers))
+
+    def load_config(self):
+        config_dir = os.path.expanduser('~/.ralph-dhcp-agent')
+        if not os.path.exists(config_dir):
+            os.makedirs(config_dir)
+        config_file = os.path.join(config_dir, 'config.db')
+        self._config = gdbm.open(config_file, 'c')
+
+    def save_config(self):
+        self._config.close()
+
+    def get_url_hash(self, url):
+        hash_url = hashlib.md5()
+        hash_url.update(url.encode())
+        return hash_url.hexdigest()
+
+    def get_last_modified(self, url):
+        h = self.get_url_hash(url)
+        try:
+            last = self._config[CONFIG_HTTP_LAST_MODIFIED + h]
+        except KeyError:
+            last = None
+        return last
+
+    def set_last_modified(self, url, value):
+        h = self.get_url_hash(url)
+        self._config[CONFIG_HTTP_LAST_MODIFIED + h] = value
+
+    def fetch_configuration(self, mode, request_params):
```

Comment:
docstring is missing - and the function name is not the best  - get vs fetch are simmilar, but the latter sets the new configuraton as well ( 2 operations in one)